File: frontend/src/api/membership.ts
import { request } from './client'

export type MembershipTier = 'none' | 'vip' | 'svip'
export interface Membership {
  user_id: string
  tier: MembershipTier
  expires_at: string | null
  active: boolean
  owner_limit_override: number | null
  owner_limit: number
  owned_plans: number
  revision: number
  source: 'none' | 'rollout' | 'admin' | 'payment'
  billing_started: boolean
  payment_enabled: boolean
}
export interface MembershipOrder {
  id: string
  user_id: string
  product: 'vip' | 'svip' | 'upgrade'
  amount_cents: number
  upgrade_expires_at: string | null
  payment_method: 'alipay' | 'wxpay'
  status: 'pending' | 'expired' | 'paid' | 'review_required'
  trade_no: string | null
  created_at: string
  expires_at: string
  paid_at: string | null
  service_expires_at: string | null
}
export interface MembershipAdjustment {
  tier: MembershipTier
  expires_at: string | null
  owner_limit_override: number | null
  revision: number
  reason: string
  review_order_id: string
}
const root = (userID: string) => userID ? `/api/admin/users/${userID}/membership` : '/api/membership'
export const membershipAPI = {
  get: (userID = '') => request<Membership>(root(userID)),
  orders: (userID = '') => request<MembershipOrder[]>(`${root(userID)}/orders`),
  checkout: (product: MembershipOrder['product'], method: MembershipOrder['payment_method']) => request<{ order: MembershipOrder; pay_url: string }>('/api/membership/orders', { method: 'POST', body: JSON.stringify({ product, payment_method: method }) }),
  verify: (id: string) => request<MembershipOrder>(`/api/membership/orders/${id}/verify`, { method: 'POST' }),
  cancel: (id: string) => request<{ cancelled: boolean }>(`/api/membership/orders/${id}/cancel`, { method: 'POST' }),
  adjust: (userID: string, input: MembershipAdjustment) => request<{ updated: boolean }>(root(userID), { method: 'PATCH', body: JSON.stringify(input) }),
}

File: backend/internal/domain/membership.go
package domain

import "time"

type Membership struct {
	UserID             string     `json:"user_id"`
	Tier               string     `json:"tier"`
	ExpiresAt          *time.Time `json:"expires_at"`
	Active             bool       `json:"active"`
	OwnerLimitOverride *int       `json:"owner_limit_override"`
	OwnerLimit         int        `json:"owner_limit"`
	OwnedPlans         int        `json:"owned_plans"`
	Revision           int64      `json:"revision"`
	Source             string     `json:"source"`
	BillingStarted     bool       `json:"billing_started"`
	PaymentEnabled     bool       `json:"payment_enabled"`
}

type MembershipOrder struct {
	ID                    string     `json:"id"`
	UserID                string     `json:"user_id"`
	Product               string     `json:"product"`
	AmountCents           int        `json:"amount_cents"`
	MembershipRevision    int64      `json:"-"`
	UpgradeExpiresAt      *time.Time `json:"upgrade_expires_at"`
	ProviderPID           string     `json:"-"`
	ProviderBaseURL       string     `json:"-"`
	ProviderKeyCiphertext []byte     `json:"-"`
	PaymentMethod         string     `json:"payment_method"`
	Status                string     `json:"status"`
	TradeNo               *string    `json:"trade_no"`
	CreatedAt             time.Time  `json:"created_at"`
	ExpiresAt             time.Time  `json:"expires_at"`
	PaidAt                *time.Time `json:"paid_at"`
	ServiceExpiresAt      *time.Time `json:"service_expires_at"`
}

type MembershipAdjustment struct {
	Tier               string     `json:"tier"`
	ExpiresAt          *time.Time `json:"expires_at"`
	OwnerLimitOverride *int       `json:"owner_limit_override"`
	Revision           int64      `json:"revision"`
	Reason             string     `json:"reason"`
	ReviewOrderID      string     `json:"review_order_id"`
}
